mea/src/mpsc/mod.rs

```diff
@@ -0,0 +1,310 @@
+    /// use mea::mpsc::TryRecvError;
+    /// let (tx, mut rx) = mpsc::unbounded();
+    ///
+    /// tx.send("hello").unwrap();
+    ///
+    /// assert_eq!(Ok("hello"), rx.try_recv());
+    /// assert_eq!(Err(TryRecvError::Empty), rx.try_recv());
+    ///
+    /// tx.send("hello").unwrap();
+    /// drop(tx);
+    ///
+    /// assert_eq!(Ok("hello"), rx.try_recv());
+    /// assert_eq!(Err(TryRecvError::Disconnected), rx.try_recv());
+    /// # }
+    /// ```
+    pub fn try_recv(&mut self) -> Result<T, TryRecvError> {
+        match self.receiver.try_recv() {
+            Ok(v) => Ok(v),
+            Err(std::sync::mpsc::TryRecvError::Disconnected) => Err(TryRecvError::Disconnected),
+            Err(std::sync::mpsc::TryRecvError::Empty) => Err(TryRecvError::Empty),
+        }
+    }
+
+    /// Receives the next value for this receiver.
+    ///
+    /// This method returns `None` if the channel has been closed and there are
+    /// no remaining messages in the channel's buffer. This indicates that no
+    /// further values can ever be received from this `Receiver`. The channel is
+    /// closed when all senders have been dropped.
+    ///
+    /// If there are no messages in the channel's buffer, but the channel has
+    /// not yet been closed, this method will sleep until a message is sent or
+    /// the channel is closed.
+    ///
+    /// # Cancel safety
+    ///
+    /// This method is cancel safe. If `recv` is used as the event in a `select` statement
+    /// and some other branch completes first, it is guaranteed that no messages were received
+    /// on this channel.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #[tokio::main]
+    /// # async fn main() {
+    /// use mea::mpsc;
+    /// let (tx, mut rx) = mpsc::unbounded();
+    ///
+    /// tokio::spawn(async move {
+    ///     tx.send("hello").unwrap();
+    /// });
+    ///
+    /// assert_eq!(Some("hello"), rx.recv().await);
+    /// assert_eq!(None, rx.recv().await);
+    /// # }
+    /// ```
+    ///
+    /// Values are buffered:
+    ///
+    /// ```
+    /// # #[tokio::main]
+    /// # async fn main() {
+    /// use mea::mpsc;
+    /// let (tx, mut rx) = mpsc::unbounded();
+    ///
+    /// tx.send("hello").unwrap();
+    /// tx.send("world").unwrap();
+    ///
+    /// assert_eq!(Some("hello"), rx.recv().await);
+    /// assert_eq!(Some("world"), rx.recv().await);
+    /// # }
+    /// ```
+    pub async fn recv(&mut self) -> Option<T> {
+        poll_fn(|cx| self.poll_recv(cx)).await
+    }
+
+    fn poll_recv(&mut self, cx: &mut Context<'_>) -> Poll<Option<T>> {
+        match self.receiver.try_recv() {
+            Ok(v) => Poll::Ready(Some(v)),
+            Err(std::sync::mpsc::TryRecvError::Disconnected) => Poll::Ready(None),
+            Err(std::sync::mpsc::TryRecvError::Empty) => {
+                let waker = Box::into_raw(Box::new(cx.waker().clone()));
+                self.state.rx_task.store(waker, Ordering::Release);
```

Comment:
If `poll_recv` was previously called, or the `Future` returned by `recv` was dropped, and no `send` has been performed since then, then `rx_task` will contain the `Waker` from that previous invocation and this `store` will leak both the `Waker` and the `Box` it was stored in.

Reply:
Yeah.

Not sure if add a drop glue within `UnboundedReceiver`'s Drop impl would help. Let me think ..

Reply:
Use `AtomicOptionBox` should be better. Updating ...

Reply:
> Not sure if add a drop glue within UnboundedReceiver's Drop impl would help

No, since you can call `poll_recv` or `recv` multiple times without dropping `UnboundedReceiver`. Adding drop glue to the return value of `recv` could help, but it would solve only half the issue, so I wouldn't recommend it.

---

The way I would fix it would be to use `swap` instead of `store`, and drop the `Waker` and `Box` that was stored in `rx_task` (if not null).